### ISSUE TITLE
What is the reason to append column of 1s to X_train during pre-processing?

### ISSUE THREAD

**Issue by nirdesh80** [ushasi/MachineLearning#2]
Hi @ushasi ,
I was going through the code for LInearRegression_multicollinearity_preprocessing.ipynb ( [https://github.com/ushasi/MachineLearning/blob/main/Linear%20Regression/LInearRegression_multicollinearity_preprocessing.ipynb](url))
In here during pre-processing (code block # 9) just after normalizing the train-validate data, a column of 1s is concatenated to X_train, x_val, and later to X_test
Could you please explain the reason for this concatenation?

`        X_train = np.c_[np.ones(len(X_train),dtype='int64'), X_train.to_numpy()]    # append 1s to the features in train set
        X_val = np.c_[np.ones(len(X_val),dtype='int64'), X_val.to_numpy()]  # append 1s to the features in validation set
`

Thanks,
Niranjan

**Comment by parth-shettiwar** [comment]
Hi @nirdesh80 ,
Column of ones is attached to account for the bias. To give an intuition:
In linear regression we compute output Y = WX + b, where W are the weights, X is the input feature vector, b is the bias. Here both W and b are trainable. Say X is a feature vector of length (10 x 1), then W will  be vector of length (1x10), and b will of length 1 .We can rewrite above equation as Y = WX + b * 1. Since coefficient of b is always 1, hence we can simply append 1 to X vector and absorb b in W vector (since both b and W are trainable). Hence Y = W"X", where W" = vector of length (1 x 11), since b is absorbed and X" = X with column of one appended, making it 11 x 1 vector.
Hope this clears it out.

**Comment by nirdesh80** [comment]
> Hi @nirdesh80 ,
> Column of ones is attached to account for the bias. To give an intuition:
> In linear regression we compute output Y = WX + b, where W are the weights, X is the input feature vector, b is the bias. Here both W and b are trainable. Say X is a feature vector of length (10 x 1), then W will be vector of length (1x10), and b will of length 1 .We can rewrite above equation as Y = WX + b * 1. Since coefficient of b is always 1, hence we can simply append 1 to X vector and absorb b in W vector (since both b and W are trainable). Hence Y = W"X", where W" = vector of length (1 x 11), since b is absorbed and X" = X with column of one appended, making it 11 x 1 vector.
> Hope this clears it out.

Yes this makes sense, just that maybe  you meant, X'' would be (10 x **2**) , an added column of ones?
then y^ = row-wise sum of element-wise _multiplication_ of W and X 
OR 
y^ = _dot_ product of W and X.T
 
but then this is required when one plans to implement Linear Regression from scratch right?
since we are using the fit function from sklearn in this code was this appending required?

**Comment by ushasi** [comment]
Yes, you are right. This particular code was just a modification of the previous code which was implemented from scratch to show the multi-collinearity module. Since the remainder of the code uses the standard libraries, you can skip this appending one pre-processing step. It is redundant.
